pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java

```diff
@@ -292,6 +249,69 @@ public void addFailed(ManagedLedgerException exception, Object ctx) {
         return completableFuture;
     }
 
+    private void handleMetadataEntry(PositionImpl logPosition, PendingAckMetadataEntry pendingAckMetadataEntry) {
+        // store the persistent position in to memory
+        // store the max position of this entry retain
+        if (pendingAckMetadataEntry.getPendingAckOp() != PendingAckOp.ABORT
+                && pendingAckMetadataEntry.getPendingAckOp() != PendingAckOp.COMMIT) {
+            Optional<PendingAckMetadata> optional = pendingAckMetadataEntry.getPendingAckMetadatasList()
+                    .stream().max((o1, o2) -> ComparisonChain.start().compare(o1.getLedgerId(),
+                            o2.getLedgerId()).compare(o1.getEntryId(), o2.getEntryId()).result());
+
+            optional.ifPresent(pendingAckMetadata -> {
+                PositionImpl nowPosition = PositionImpl.get(pendingAckMetadata.getLedgerId(),
+                        pendingAckMetadata.getEntryId());
+
+                if (nowPosition.compareTo(maxAckPosition) > 0) {
+                    maxAckPosition = nowPosition;
+                }
+                if (logAppendTimes.get() > upperLimitOfLogAppendTimes) {
```

Comment:
we should use `currentIndexLag` and `maxIndexLag` here to keep consistent with the `LogIndexLagBackoff `

Reply:
Also here also have a race condition, if more than 1 threads reach line 269, it will not respect the constraints of backoff

Reply:
Sorry, I can not understand you mind. In this proposal, we use the size of pendingAckLogIndex to get  the next backoff.
And we do not implement the maxIndexLag now.

Reply:
And the execution here is synchronous, there is no race

Reply:
> Sorry, I can not understand you mind. In this proposal, we use the size of pendingAckLogIndex to get the next backoff.

I mean we already introduced `Lag` concept here, why need to introduce another concept named `*AppendTimes`, from the current implementation, logAppendTimes == currentIndexLag, upperLimitOfLogAppendTimes == maxIndexLag right? just a different name, but the new name makes the code more cumbersome to read

We use a LagBackoff to calculate the next maxIndexLag, more entries are added to the entry log, the currentIndexLog increases, if reach the maxIndexLag, to get the next maxIndexLag and reset the currentIndexLag. 

Or we can avoid resetting the `currentIndexLag`, just keep it increasing and use a `nextIndexLag`. If reach the `currentIndexLag` == `nextIndexLag`, reset the `nextIndexLag`

Reply:
> And the execution here is synchronous, there is no race

If the execution here is synchronous, why need to use `AtomicLong`?

Reply:
>We use a LagBackoff to calculate the next maxIndexLag, more entries are added to the entry log, the currentIndexLog increases, if reach the maxIndexLag, to get the next maxIndexLag and reset the currentIndexLag.
>Or we can avoid resetting the currentIndexLag, just keep it increasing and use a nextIndexLag. If reach the currentIndexLag == nextIndexLag, reset the nextIndexLag.
The first implementation is a bit more readable than the second, I tend to use the first implementation.

Reply:
>If the execution here is synchronous, why need to use AtomicLong?
for memory visibility